### ISSUE TITLE
TypeError: null is not an object (evaluating '_qz.websocket.connection.onerror')

### ISSUE THREAD

**Issue by ScutelnicuIonel** [qzind/tray#261]
This is the line that is causing the problem: https://github.com/qzind/tray/blob/3afd7f29e0d735324aff3b4322e3d30ae6d54d11/js/qz-tray.js#L154

Reported JS error:
```
{
  "message": "TypeError: null is not an object (evaluating '_qz.websocket.connection.onerror')",
  "line": 154,
  "app_name": "Netscape",
  "app_version": "5.0 (Macintosh; Intel Mac OS X 10_13_1) AppleWebKit/604.3.5 (KHTML, like Gecko) Version/11.0.1 Safari/604.3.5",
  "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_1) AppleWebKit/604.3.5 (KHTML, like Gecko) Version/11.0.1 Safari/604.3.5",
  "platform": "MacIntel",
  "cookie_enabled": true
}
```

The problem seems to be from a Safari 11 on Mac OS X 10 (user agent decoder) i don`t have a mac so i cannot help you with tests.

**Comment by tresf** [comment]
@ScutelnicuIonel is this separate from #260?

**Comment by ScutelnicuIonel** [comment]
@tresf Yes.

**Comment by tresf** [comment]
@ScutelnicuIonel we have various MacOS environments, so we would be happy to reproduce. Can you please provide the steps to reproduce?  We test each release on Safari, so we would need a bit more go to on.

![image](https://user-images.githubusercontent.com/6345473/33997128-68ae8ace-e0b1-11e7-8f3b-4271928413d7.png)

**Comment by ScutelnicuIonel** [comment]
Sorry but the only information that we have is that JSON, the server receive it from a reporting tool, we don`t  know what action the user has done, or in what step the error occur.

**Comment by tresf** [comment]
Something weird is definitely going on... 

I just tried to use the `qz:launch` button in sample.html and it spammed the browser with connection attempts rendering QZ Tray unusable.

![screen shot 2017-12-15 at 2 09 07 am](https://user-images.githubusercontent.com/6345473/34030930-11b019b8-e13d-11e7-88e2-b60d58f777cb.png)

I believe the failure to call `onerror()` is preventing the websocket from closing.  We'll take a closer look.

**Comment by bberenz** [comment]
So line 154 was some code for older versions of safari not actually calling `onerror` on websocket errors. The latest safari appears to have fixed that, so I added a check on the connection object in 911c525e5ad9f0042eaa63fe1158ef4104a1a013 to see if we already cleared it out. This way we can still support the older safari's for now.

As for the qz:launch bug... I don't exactly know the cause, as I couldn't replicate it in a clean codebase, but certain conditions appear to allow safari to make multiple reject calls on promises. Something that should _never_ happen according to the spec. 911c525e5ad9f0042eaa63fe1158ef4104a1a013 also adds in our own "only run once" code to handle this weird edge-case.
